### PR TITLE
OnKeySold hook

### PR DESCRIPTION
**Pull request by NickCuso**
# Description

Implements a hook for key sales which enables new features.
Anyone (person or contract) can register an implementer via ERC-1820.  The test simply logs this to confirm it was actually called.  We can follow up in the future with a more practical example of how this may be leveraged.

# Issues

Fixes #
Refs #4015

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [x] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [x] My code follows the style guidelines of this project, including naming conventions
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: smart-contracts/contracts/mixins/MixinPurchase.sol [hunk 1]
```diff
@@ -80,7 +84,10 @@ contract MixinPurchase is
     );
 
     // We explicitly allow for greater amounts of ETH to allow 'donations'
+    // Security: after state changes to minimize risk of re-entrancy
+    uint pricePaid = _chargeAtLeast(netPrice);
+
     // Security: last line to minimize risk of re-entrancy
-    _chargeAtLeast(netPrice);
+    _onKeySold(_recipient, _referrer, pricePaid, _data);
```
Comment (julien51) [inline smart-contracts/contracts/mixins/MixinPurchase.sol]: Isn't there a re-entrancy risk here if the hook calls back `purchase`?

Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinPurchase.sol]: It is possible to reenter via the hook.  However that would be fine.

Reentrancy becomes problematic when some state changes for the transaction are not complete before the external call is made.  Here we were able to make it the last line of the purchase function which makes this clearly safe.  Any reentrancy would be on-par with simply calling `purchase` again in a future tx.

**Comment by NickCuso** [comment]
> Do you mind checking if this does not go over the gas limit upgrade unlock?

Deploy: 6.1 mil or 91.9% of the limit 
(available from the gas report at the end of the smart contract tests in CI)

**Comment by julien51** [comment]
O, I really think we need to prioritize upgrading Unlock *without* changing the ABI for PublicLock.

**Comment by NickCuso** [comment]
...I thought we discussed this and were okay breaking ABI.

`purchase` is already different from the current PublicLock..

File: smart-contracts/contracts/mixins/MixinPurchase.sol (hunk 1 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinPurchase.sol]: Yup, agreed that this looks fine as it follows the "checks, effects, interactions" pattern.

**Pull request merged**